File: ImageProcessor.m
//
//  ImageProcessor.m
//  Tutu
//
//  图像处理工具类实现
//

#import "ImageProcessor.h"
#import <CoreImage/CoreImage.h>

@implementation ImageProcessor

#pragma mark - 滤镜应用

+ (UIImage *)applyFilter:(UIImage *)image withType:(TTFilterType)filterType {
    if (!image) return nil;
    
    // 如果是原图，直接返回
    if (filterType == TTFilterTypeNone) {
        return image;
    }
    
    // 创建CIContext和CIImage
    CIContext *context = [CIContext contextWithOptions:nil];
    CIImage *inputImage = [[CIImage alloc] initWithImage:image];
    
    // 根据滤镜类型选择对应的CIFilter
    CIFilter *filter = nil;
    
    switch (filterType) {
        case TTFilterTypeMono:
            filter = [CIFilter filterWithName:@"CIPhotoEffectMono"];
            break;
        case TTFilterTypeTransfer:
            filter = [CIFilter filterWithName:@"CIPhotoEffectTransfer"];
            break;
        case TTFilterTypeChrome:
            filter = [CIFilter filterWithName:@"CIPhotoEffectChrome"];
            break;
        case TTFilterTypeInstant:
            filter = [CIFilter filterWithName:@"CIPhotoEffectInstant"];
            break;
        case TTFilterTypeFade:
            filter = [CIFilter filterWithName:@"CIPhotoEffectFade"];
            break;
        default:
            return image;
    }
    
    // 设置输入图像
    [filter setValue:inputImage forKey:kCIInputImageKey];
    
    // 获取输出图像
    CIImage *outputImage = filter.outputImage;
    if (!outputImage) return image;
    
    // 转换为UIImage
    CGImageRef cgImage = [context createCGImage:outputImage fromRect:outputImage.extent];
    UIImage *resultImage = [UIImage imageWithCGImage:cgImage];
    CGImageRelease(cgImage);
    
    return resultImage;
}

#pragma mark - 图像调整

+ (UIImage *)adjustImage:(UIImage *)image 
              brightness:(CGFloat)brightness 
               contrast:(CGFloat)contrast 
              saturation:(CGFloat)saturation {
    if (!image) return nil;
    
    // 如果所有参数都是默认值，直接返回
    if (brightness == 0 && contrast == 1.0 && saturation == 1.0) {
        return image;
    }
    
    // 创建CIContext和CIImage
    CIContext *context = [CIContext contextWithOptions:nil];
    CIImage *inputImage = [[CIImage alloc] initWithImage:image];
    
    // 创建颜色控制滤镜
    CIFilter *filter = [CIFilter filterWithName:@"CIColorControls"];
    [filter setValue:inputImage forKey:kCIInputImageKey];
    [filter setValue:@(brightness) forKey:kCIInputBrightnessKey];
    [filter setValue:@(contrast) forKey:kCIInputContrastKey];
    [filter setValue:@(saturation) forKey:kCIInputSaturationKey];
    
    // 获取输出图像
    CIImage *outputImage = filter.outputImage;
    if (!outputImage) return image;
    
    // 转换为UIImage
    CGImageRef cgImage = [context createCGImage:outputImage fromRect:outputImage.extent];
    UIImage *resultImage = [UIImage imageWithCGImage:cgImage];
    CGImageRelease(cgImage);
    
    return resultImage;
}

#pragma mark - 图像变换

+ (UIImage *)rotateImage:(UIImage *)image degrees:(CGFloat)degrees {
    if (!image) return nil;
    
    // 转换角度为弧度
    CGFloat radians = degrees * M_PI / 180.0;
    
    // 获取图像尺寸
    CGSize size = image.size;
    
    // 计算旋转后的尺寸
    CGFloat newWidth = fabsf(size.width * cosf(radians)) + fabsf(size.height * sinf(radians));
    CGFloat newHeight = fabsf(size.height * cosf(radians)) + fabsf(size.width * sinf(radians));
    CGSize newSize = CGSizeMake(newWidth, newHeight);
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(newSize, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 移动原点到中心
    CGContextTranslateCTM(context, newSize.width / 2, newSize.height / 2);
    
    // 应用旋转变换
    CGContextRotateCTM(context, radians);
    
    // 绘制原图
    [image drawInRect:CGRectMake(-size.width / 2, -size.height / 2, size.width, size.height)];
    
    // 获取旋转后的图像
    UIImage *rotatedImage = UIGraphicsGetImageFromCurrentImageContext();
    
    // 结束图形上下文
    UIGraphicsEndImageContext();
    
    return rotatedImage;
}

+ (UIImage *)scaleImage:(UIImage *)image byScale:(CGFloat)scale {
    if (!image || scale <= 0) return nil;
    
    // 计算新尺寸
    CGSize newSize = CGSizeMake(image.size.width * scale, image.size.height * scale);
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(newSize, NO, image.scale);
    
    // 绘制缩放后的图像
    [image drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
    
    // 获取缩放后的图像
    UIImage *scaledImage = UIGraphicsGetImageFromCurrentImageContext();
    
    // 结束图形上下文
    UIGraphicsEndImageContext();
    
    return scaledImage;
}

+ (UIImage *)cropImage:(UIImage *)image toRect:(CGRect)rect {
    if (!image) return nil;
    
    // 确保裁剪区域有效
    CGRect validRect = CGRectIntersection(rect, CGRectMake(0, 0, image.size.width, image.size.height));
    if (CGRectIsNull(validRect)) return nil;
    
    // 考虑Retina屏幕的缩放
    CGRect scaledRect = CGRectMake(validRect.origin.x * image.scale,
                                 validRect.origin.y * image.scale,
                                 validRect.size.width * image.scale,
                                 validRect.size.height * image.scale);
    
    // 获取CGImage并裁剪
    CGImageRef originalCGImage = image.CGImage;
    CGImageRef croppedCGImage = CGImageCreateWithImageInRect(originalCGImage, scaledRect);
    
    // 创建新的UIImage
    UIImage *croppedImage = [UIImage imageWithCGImage:croppedCGImage scale:image.scale orientation:image.imageOrientation];
    
    // 释放资源
    CGImageRelease(croppedCGImage);
    
    return croppedImage;
}

#pragma mark - 组合效果

+ (UIImage *)applyCombinedEffects:(UIImage *)image
                       filterType:(TTFilterType)filterType
                        brightness:(CGFloat)brightness
                         contrast:(CGFloat)contrast
                        saturation:(CGFloat)saturation {
    if (!image) return nil;
    
    // 先应用亮度、对比度和饱和度调整
    UIImage *adjustedImage = [self adjustImage:image 
                                    brightness:brightness 
                                     contrast:contrast 
                                    saturation:saturation];
    
    // 再应用滤镜效果
    UIImage *finalImage = [self applyFilter:adjustedImage withType:filterType];
    
    return finalImage;
}

#pragma mark - 工具方法

+ (NSString *)nameForFilterType:(TTFilterType)filterType {
    switch (filterType) {
        case TTFilterTypeNone:
            return @"原图";
        case TTFilterTypeMono:
            return @"黑白";
        case TTFilterTypeTransfer:
            return @"怀旧";
        case TTFilterTypeChrome:
            return @"冷色调";
        case TTFilterTypeInstant:
            return @"暖色调";
        case TTFilterTypeFade:
            return @"复古";
        default:
            return @"未知";
    }
}

+ (NSArray<NSNumber *> *)allFilterTypes {
    return @[@(TTFilterTypeNone), 
             @(TTFilterTypeMono), 
             @(TTFilterTypeTransfer), 
             @(TTFilterTypeChrome), 
             @(TTFilterTypeInstant), 
             @(TTFilterTypeFade)];
}

@end

File: ImageProcessor.cpp
#include "ImageProcessor.hpp"
#import <CoreGraphics/CoreGraphics.h>
#import <Accelerate/Accelerate.h>

UIImage* ImageProcessor::applyBasicFilter(UIImage* image, const std::string& filterType) {
    if (!image) return nullptr;
    
    // 获取图像尺寸
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(size, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 绘制原始图像
    [image drawInRect:rect];
    
    // 根据滤镜类型应用效果
    if (filterType == "grayscale") {
        // 灰度滤镜
        CGContextSetBlendMode(context, kCGBlendModeSaturation);
        CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0);
        CGContextFillRect(context, rect);
    } else if (filterType == "sepia") {
        // 棕褐滤镜
        CGContextSetBlendMode(context, kCGBlendModeColor);
        CGContextSetRGBFillColor(context, 0.8, 0.6, 0.4, 0.3);
        CGContextFillRect(context, rect);
    }
    
    // 获取处理后的图像
    UIImage* filteredImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return filteredImage;
}

UIImage* ImageProcessor::rotateImage(UIImage* image, float degrees) {
    if (!image) return nullptr;
    
    // 转换角度为弧度
    CGFloat radians = degrees * M_PI / 180.0;
    
    // 计算旋转后的尺寸
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(size, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 移动到中心点并旋转
    CGContextTranslateCTM(context, size.width / 2, size.height / 2);
    CGContextRotateCTM(context, radians);
    CGContextTranslateCTM(context, -size.width / 2, -size.height / 2);
    
    // 绘制图像
    [image drawInRect:rect];
    
    // 获取旋转后的图像
    UIImage* rotatedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return rotatedImage;
}

UIImage* ImageProcessor::scaleImage(UIImage* image, float scale) {
    if (!image) return nullptr;
    
    // 计算缩放后的尺寸
    CGSize originalSize = image.size;
    CGSize newSize = CGSizeMake(originalSize.width * scale, originalSize.height * scale);
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(newSize, NO, image.scale);
    CGRect rect = CGRectMake(0, 0, newSize.width, newSize.height);
    
    // 绘制缩放后的图像
    [image drawInRect:rect];
    
    // 获取缩放后的图像
    UIImage* scaledImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return scaledImage;
}

UIImage* ImageProcessor::cropImage(UIImage* image, CGRect cropRect) {
    if (!image) return nullptr;
    
    // 创建图形上下文
    UIGraphicsBeginImageContextWithOptions(cropRect.size, NO, image.scale);
    
    // 绘制裁剪区域
    [image drawInRect:CGRectMake(-cropRect.origin.x, -cropRect.origin.y, image.size.width, image.size.height)];
    
    // 获取裁剪后的图像
    UIImage* croppedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return croppedImage;
}

UIImage* ImageProcessor::adjustBrightness(UIImage* image, float brightness) {
    if (!image) return nullptr;
    
    // 调整亮度的简单实现
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    UIGraphicsBeginImageContextWithOptions(size, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 绘制原始图像
    [image drawInRect:rect];
    
    // 应用亮度调整
    CGContextSetBlendMode(context, kCGBlendModeOverlay);
    CGContextSetRGBFillColor(context, brightness, brightness, brightness, 0.5);
    CGContextFillRect(context, rect);
    
    UIImage* adjustedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return adjustedImage;
}

UIImage* ImageProcessor::adjustContrast(UIImage* image, float contrast) {
    if (!image) return nullptr;
    
    // 调整对比度的简单实现
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    UIGraphicsBeginImageContextWithOptions(size, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 绘制原始图像
    [image drawInRect:rect];
    
    // 应用对比度调整
    CGContextSetBlendMode(context, kCGBlendModeSoftLight);
    CGContextSetRGBFillColor(context, contrast, contrast, contrast, 0.3);
    CGContextFillRect(context, rect);
    
    UIImage* adjustedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return adjustedImage;
}

UIImage* ImageProcessor::adjustSaturation(UIImage* image, float saturation) {
    if (!image) return nullptr;
    
    // 调整饱和度的简单实现
    CGSize size = image.size;
    CGRect rect = CGRectMake(0, 0, size.width, size.height);
    
    UIGraphicsBeginImageContextWithOptions(size, NO, image.scale);
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // 绘制原始图像
    [image drawInRect:rect];
    
    // 应用饱和度调整
    if (saturation < 1.0) {
        CGContextSetBlendMode(context, kCGBlendModeSaturation);
        CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0 - saturation);
        CGContextFillRect(context, rect);
    }
    
    UIImage* adjustedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return adjustedImage;
}

void ImageProcessor::processImagePixels(void* pixelData, int width, int height, const std::string& operation) {
    // 像素级图像处理的基础实现
    // 这里可以添加更复杂的图像处理算法
    
    unsigned char* pixels = static_cast<unsigned char*>(pixelData);
    int pixelCount = width * height * 4; // RGBA
    
    if (operation == "invert") {
        // 颜色反转
        for (int i = 0; i < pixelCount; i += 4) {
            pixels[i] = 255 - pixels[i];     // R
            pixels[i + 1] = 255 - pixels[i + 1]; // G
            pixels[i + 2] = 255 - pixels[i + 2]; // B
            // Alpha 通道保持不变
        }
    }
}